reusable_components_nlp/README.md

```diff
@@ -0,0 +1,989 @@
+
+# End-to-end Reusable ML Pipeline with Seldon and Kubeflow
+
+In this example we showcase how to build re-usable components to build an ML pipeline that can be trained and deployed at scale.
+
+We will automate content moderation on the Reddit comments in /r/science building a machine learning NLP model with the following components:
+
+![](img/completed-pipeline-deploy.jpg)
+
+This tutorial will break down in the following sections:
+
+1) Run all the services (Kubeflow and Seldon)
+
+2) Test and build all our reusable pipeline steps
+
+3) Use Kubeflow to Train the Pipeline and Deploy to Seldon
+
+5) Test Seldon Deployed ML REST Endpoints
+
+6) Visualise Seldon's Production ML Pipelines
+
+# Before you start
+Make sure you install the following dependencies, as they are critical for this example to work:
+
+* Helm v2.13.1+
+* A Kubernetes cluster running v1.13 or above (minkube / docker-for-windows work well if enough RAM)
+* kubectl v1.14+
+* ksonnet v0.13.1+
+* kfctl 0.5.1 - Please use this exact version as there are major changes every few months
+* Python 3.6+
+* Python DEV requirements (we'll install them below)
+
+Let's get started! 🚀🔥 We will be building the end-to-end pipeline below:
+
+![](img/kubeflow-seldon-nlp-full.jpg)
+
+
+
+```python
+!cat requirements-dev.txt
+```
+
+    python-dateutil
+    https://storage.googleapis.com/ml-pipeline/release/0.1.20/kfp.tar.gz
+    kubernetes
+    click
+    seldon_core
+    numpy
+
+
+
+```python
+!pip install -r requirements-dev.txt
+```
+
+## 1) Run all the services (Kubeflow and Seldon)
+Kubeflow's CLI allows us to create a project which will allow us to build the configuration we need to deploy our kubeflow and seldon clusters.
+
+
+```python
+!kfctl init kubeflow-seldon
+!ls kubeflow-seldon
+```
+
+    app.yaml
+
+
+Now we run the following commands to basically launch our Kubeflow cluster with all its components. 
+
+It may take a while to download all the images for Kubeflow so feel free to make yourself a cup of ☕.
+
+If you have a terminal you can see how the containers are created in real-time by running `kubectl get pods -n kubeflow -w`.
+
+
+```bash
+%%bash
+cd kubeflow-seldon
+kfctl generate all -V
+kfctl apply all -V
+```
+
+### Now let's run Seldon 
+For this we'll need Helm to be running, so we'll initialise it.
+
+
+```bash
+%%bash
+helm init 
```

Comment:
Rather than introduce another packaging tool, let's use kustomize here (ksonnet if necessary, but since it's being retired I'd rather not use it for net-new code).

Reply:
I see where you're coming from. At seldon we use helm as our default package manager tho, and the latest seldon charts are continuously updated in the helm repo. There is currently a deployment of Kubeflow that has a version of Seldon but the version is quite old and would need updating for that. What would you suggest here?

Reply:
This is a common pain point & we haven't worked out a great way to accommodate it yet. The motivation for this request is to reduce the user requirement for additional tooling. The other point you brought up is that KF currently ships an out-of-date version of Seldon, which isn't what we want either. I think it's worthwhile to update the component that ships with KF releases. I'm unopinionated on which occurs first, but the same commands should be applicable both here and there.

Reply:
Also, I know this has been a blocker for integration of other components as well, so it's worth bringing up in slack or on a weekly call so that we can figure out the best approach.